contracts/compressor/IntentCompressor.sol

```diff
@@ -0,0 +1,336 @@
+    uint8 destinationChainIndex;
+    // Reward token
+    uint8 rewardTokenIndex;
+    uint48 rewardAmount;
+    // Route token
+    uint8 routeTokenIndex;
+    uint48 routeAmount;
+    // Expiry duration
+    uint24 expiryDuration;
+    // Salt
+    bytes8 salt;
+}
+
+struct EncodedFulfillment {
+    uint8 sourceChainIndex;
+    uint8 destinationChainIndex;
+    // Reward token
+    uint8 routeTokenIndex;
+    uint48 routeAmount;
+    // Prove type (INSTANT = 0, BATCH = 1)
+    uint8 proveType;
+    // Claimant
+    address claimant;
+}
+
+contract IntentCompressor {
+    using SafeERC20 for IERC20;
+
+    /**
+     * @notice Thrown when the vault has insufficient token allowance for reward funding
+     * @param token The token address
+     * @param spender The spender address
+     * @param amount The amount of tokens required
+     */
+    error InsufficientTokenAllowance(
+        address token,
+        address spender,
+        uint256 amount
+    );
+
+    address public immutable PROVER;
+    IInbox public immutable INBOX;
+    IIntentSource public immutable INTENT_SOURCE;
+
+    constructor(address _intentSource, address _inbox, address _prover) {
+        INBOX = IInbox(_inbox);
+        INTENT_SOURCE = IIntentSource(_intentSource);
+        PROVER = _prover;
+    }
+
+    function fulfill(
+        bytes32 payload,
+        bytes32 rewardHash,
+        bytes32 routeSalt
+    ) external returns (bytes[] memory) {
+        EncodedFulfillment memory encodedFulfillment = decodeFulfillPayload(
+            payload
+        );
+
+        Route memory route = _constructRoute(
+            routeSalt,
+            encodedFulfillment.sourceChainIndex,
+            encodedFulfillment.destinationChainIndex,
+            encodedFulfillment.routeTokenIndex,
+            encodedFulfillment.routeAmount
+        );
+
+        bytes32 routeHash = keccak256(abi.encode(route));
+        bytes32 intentHash = keccak256(abi.encodePacked(routeHash, rewardHash));
+
+        require(
+            route.tokens.length == 1,
+            "Cannot fulfill intent multiple tokens"
+        );
+
+        // Approve route token
+        TokenAmount memory routeToken = route.tokens[0];
+        IERC20(routeToken.token).approve(address(INBOX), routeToken.amount);
+
+        if (encodedFulfillment.proveType == 1) {
+            return
+                INBOX.fulfillHyperBatched(
+                    route,
+                    rewardHash,
+                    encodedFulfillment.claimant,
+                    intentHash,
+                    PROVER
+                );
+        } else {
+            return
+                INBOX.fulfillHyperInstant(
+                    route,
+                    rewardHash,
+                    encodedFulfillment.claimant,
+                    intentHash,
+                    PROVER
+                );
+        }
+    }
+
+    function publishTransferIntentAndFund(
+        bytes32 payload
+    ) external returns (bytes32 intentHash) {
```

Comment:
Could be payable and native value to the reward and route call for native transfer? Or should we use `tokenIndex == 0` for a native token instead of USDC?